### PR TITLE
Definition MetaStructure

### LINKED ISSUE #479: I want the AaC definition structure to have a node-based structure
As an AaC developer and validator developer, I want the AaC definition structure to use a node structure instead of dictionaries, so that I can add metadata and object references to the structural nodes/values in an AaC definition.

The goal of this story is to create a parallel definition structure to the straight yaml-parsed `dict`s and `list`s. This alternate structure is intended to pro-actively calculate and collate information for the definition's structure into easily traversable nodes in a tree structure with bi-directional edges (e.g. you can traverse to both parent and children nodes). Attaching metadata about these components will help to avoid a lot of overhead associated with the current approach of computing this kind of metadata on demand -- which quickly tanks performance with more and more definitions in the context.

## Example
An example decomposition of a simple model into structural nodes would look something akin to the following.

Example model:
```yaml
model:
  name: Clock
  description: A simple clock that keeps track of the current time
  behavior:
    - name: getTime
      type: timer
      description: Get the current time
      output:
        - name: currentTime
          type: Timestamp
```

Example node structure:

root key node (top-level, root node):
```
token: 'model'
schema_definition: 'model' <-- This is a link to the schema named `model` in the core spec
node_type: 'ROOT_KEY'
children: [node('name'), node('description'), node('behavior')]
parent: None
```

schema node ('name'):
```
token: 'name'
schema_definition: 'Field' 
node_type: 'PRIMITIVE_STRUCTURE' <-- Even though primitives are enums, primitives have different behaviors
children: [node('Clock')]
parent: node('model')
```

primitive node ('Clock'):
```
token: 'Clock'
schema_definition: 'Primitives'  <-- this is the definition that defines the enum values
primitive_type: string <-- This is a field unique to primitive value nodes
node_type: 'PRIMITIVE_VALUE'
children: []
parent: node('name')
```

schema node ('description'):
```
token: 'description'
schema_definition: 'Field' 
node_type: 'PRIMITIVE_STRUCTURE'
children: [node('A simple clock that keeps track of the current time')]
parent: node('model')
```

primitive node ('A simple clock that keeps track of the current time'):
```
token: 'A simple clock that keeps track of the current time'
schema_definition: 'Primitives' 
primitive_type: string 
node_type: 'PRIMITIVE_VALUE'
children: []
parent: node('description')
```

schema node ('behavior'):
```
token: 'behavior'
node_type: 'SCHEMA_STRUCTURE'
children: [node('Clock')]
parent: node('model')
```

schema node ('behaviorEntry1'):
```
token: None <-- Could alternatively be '-' since it's a list item
node_type: 'LIST_STRUCTURE'
children: [node('Clock')]
parent: node('behavior')
```

schema node ('behaviorEntry1Name'):
```
token: 'name'
schema_definition: 'Field' 
node_type: 'PRIMITIVE_STRUCTURE'
children: [node('getTime')]
parent: node('behaviorEntry1')
```

primitive node ('getTime'):
```
... Same as the other name nodes
```

schema node ('behaviorEntry1Type'):
```
token: 'type'
schema_definition: 'Field'  
node_type: 'ENUM_STRUCTURE'
children: [node('behaviorEntry1TypeTimer')]
parent: node('behaviorEntry1')
```

schema node ('behaviorEntry1TypeTimer'):
```
token: 'timer'
schema_definition: 'BehaviorType'  <-- Provides easy lookup to verify that the value 'timer' is in these enum values
node_type: 'ENUM_VALUE'
children: []
parent: node('behaviorEntry1Type')
```

... Leaving out output for some brevity, but it should follow the same pattern as behavior.

## AC:

- [x] There is a new class `StructuralNode` in the `definition` package
  - [x] Each token in the definition's structure is represented as an AaC node
  - [x] Each node has an Enum attribute describing what type the node is
    - [x] Nodes that are sub-structure (i.e. Field, Behavior) are typed as `SCHEMA_STRUCTURE` nodes
    - [x] Root-key nodes (i.e. schema, model) are typed as `ROOT_KEY`
    - [x] Enum value nodes (non-primitive enums) are typed as `ENUM_VALUE`
    - [x] Primitive-type value nodes (i.e. string, int, date) are typed as `PRIMITIVE_VALUE`
    - [x] Nodes representing lists (i.e. Fields) are typed as `LIST_STRUCTURE`
  - [x] Each node has a reference to its defining definition
    - [x] `ROOT_KEY` nodes reference their defining structure (i.e. schema, model, extension, etc)
    - [x] `SCHEMA_STRUCTURE` nodes reference their defining `schema` definition
    - [x] `ENUM_VALUE` nodes reference their defining `enum` definition
    - [x] `PRIMITIVE_VALUE` nodes reference the primitive definition
    - [x] `LIST_STRUCTURE` nodes do not reference a schema definition
  - [x] Each node has a reference to its corresponding token/lexeme
- [x] The `parser` package is updated to include an additional step to build the AaC definition structures using the new nodes
- [x] Definitions have a new attribute `meta_structure` that contains the new node-based structure
- [x] The Definition attribute `structure` is labeled as deprecated and marked for eventual removal

### PR DESCRIPTION
**Pull request by Coffee2Bits**
# Description
This PR doesn't contain user-facing changes, but rather the underlying data model for definitions within the AaC engine. Added functionality to build a tree representation for the AaC definitions with easy references to structural/schema definitions and enum/primitive types for terminal values.

# Linked Items:
Closes #479

## Type of change
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Feature improvement

### Feature Additions
Include pertinent changes as a result of adding the new feature.
- None

### Bug Fixes
- None

### Breaking Changes
- None

### Feature Improvement Changes
- None

# Checklist:
- [x] Updated documentation accordingly.
- [x] My changes generate no new warnings.
- [x] Have updated new and existing unit tests accordingly.
- [x] Linked associated items to be closed.
- [x] Have you bumped the version if there will be a release?

### CODE REVIEW

**Comment by Coffee2Bits** [comment]
For reviewers, the example model used in the unit tests has the following layout and content. This reference may make reviewing said unit tests easier.

```yaml
model:
  name: ServiceOne
  description: ''
  components: []
  behavior:
  - name: Process Data A Request
    type: request-response
    description: Process a Data A request and return a DataB response
    tags: []
    input:
    - name: in
      type: Data A
      description: ''
    output:
    - name: out
      type: DataB
      description: ''
    acceptance:
    - scenario: Receive Data A request and return DataB response
      tags: []
      given:
      - ServiceOne is running
      when:
      - ServiceOne receives a Data A request
      then:
      - ServiceOne processes the request into a DataB response
      - ServiceOne returns the DataB response
    - scenario: Receive invalid request
      tags: []
      given:
      - ServiceOne is running
      when:
      - ServiceOne receives request that isn't a Data A request
      then:
      - ServiceOne returns an error response code
  state: []

```

File: python/src/aac/plugins/first_party/gen_protobuf/gen_protobuf_impl.py [hunk 1]
```diff
@@ -201,16 +217,16 @@ def _to_template_properties_dict(
...
+        if field[DEFINITION_FIELD_TYPE] not in active_context.get_primitive_types():
+            field[DEFINITION_FIELD_TYPE] = _sanitize_string_to_pascal_case(field.get(DEFINITION_FIELD_TYPE))
 
     return {
         "message_name": name,
         "message_description": description,
         "file_type": file_type,
         "enums": enums,
-        "fields": fields,
+        DEFINITION_FIELD_FIELDS: fields,
```
Comment (Coffee2Bits) [inline python/src/aac/plugins/first_party/gen_protobuf/gen_protobuf_impl.py:229]:
```suggestion
       "fields": fields,
```

File: python/src/aac/plugins/first_party/gen_protobuf/gen_protobuf_impl.py [hunk 2]
```diff
@@ -249,27 +265,27 @@ def _get_schema_properties(interface_structures: dict[str, Definition], data_def
...
         description_as_proto_comment = _convert_description_to_protobuf_comment(proto_field_description, 1)
 
         message_fields.append(
             {
-                "name": proto_field_name,
-                "description": description_as_proto_comment,
-                "type": sanitized_proto_field_type,
+                DEFINITION_FIELD_NAME: proto_field_name,
+                DEFINITION_FIELD_DESCRIPTION: description_as_proto_comment,
+                DEFINITION_FIELD_TYPE: sanitized_proto_field_type,
```
Comment (Coffee2Bits) [inline python/src/aac/plugins/first_party/gen_protobuf/gen_protobuf_impl.py:288]:
```suggestion
               "name": proto_field_name,
               "description": description_as_proto_comment,
               "type": sanitized_proto_field_type,
```

File: python/src/aac/lang/constants.py [hunk 3]
```diff
@@ -6,38 +6,49 @@
...
-DEFINITION_FIELD_IMPORT = "import"
-DEFINITION_FIELD_INPUT = "input"
-DEFINITION_FIELD_OUTPUT = "output"
+DEFINITION_FIELD_WHEN = "when"
 
 # Core AaC Definition Names
+DEFINITION_NAME_BEHAVIOR = "Behavior"
+DEFINITION_NAME_BEHAVIOR_TYPE = "BehaviorType"
+DEFINITION_NAME_FIELD = "Field"
+DEFINITION_NAME_MODEL = "model"
```
[the remark this answers is not in the document]
Comment (Coffee2Bits) [reply python/src/aac/lang/constants.py:47]: Ah, good catch. We didn't establish a style when we built the `model` definition so its name doesn't fit the Pascal style. Given that model is in the core spec, updating the name will be a breaking change. I'd like for us to batch a couple upcoming breaking changes to reduce turbulence, one of which will address `model` not being `Model`.

File: python/src/aac/lang/definitions/structural_node.py [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+@attrs(hash=False, eq=False)
+class StructuralNode:
+    """An Architecture-as-Code definition structural node.
+
+    These nodes are used to build a traversable tree for the definition's structure as opposed to leveraging the
+        unstructured dictionaries, lists, and primitives supplied by just parsing the yaml structure.
+
+    Attributes:
+        node_type (NodeType): Enum value used to quickly distinguish between node types
+        defining_definition (Definition): The definition that defines the structure or enumeration values for the token.
```
Comment (lizzcondrey) [inline python/src/aac/lang/definitions/structural_node.py:34]: Why just `Definition` here as the type but is `Optional['Definition']` at line 40?

File: python/src/aac/lang/definitions/structural_node.py [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+        token (Optional[Lexeme]): The corresponding token within the definition's yaml structure. Optional as nodes representing entries in a list will not have tokens.
+        parent (Optional[StructuralNode]): Reference to the node's parent; null if the node is the root node.
+        children (List[StructuralNode]): Reference the the node's children; empty if there are no children.
+    """
+    node_type: NodeType = attrib(validator=validators.instance_of(NodeType))
+    defining_definition: Optional['Definition'] = attrib()  # Can't reference Definition here for validation without creating a circular dependency.
```
Comment (lizzcondrey) [inline python/src/aac/lang/definitions/structural_node.py:40]: Why is definition needing to be in quotes here for the type check?

File: python/src/aac/lang/definitions/metadata_structure.py [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+    parent_node: StructuralNode,
+    traversed_keys: list[str],
+    language_context: LanguageContext,
+) -> None:
+    """Recursively builds the meta structure nodes and attaches them to the parent node."""
+
+    child_nodes = []
+
+    def _create_node(
+        node_type: NodeType, node_token: Optional[Lexeme] = None, defining_definition: Optional[Definition] = None
```
Comment (lizzcondrey) [inline python/src/aac/lang/definitions/metadata_structure.py:47]: Would follow similar style as above and have each parameter be on a line since there are multiple, especially since it is already split from the method name.

File: python/src/aac/io/parser/_parse_source.py [hunk 7]
```diff
@@ -104,7 +104,7 @@ def is_token_between_locations(token, inclusive_line_start: int, inclusive_line_
 
         end_of_file_offset = 1 if isinstance(end_doc_token, StreamEndToken) else 0
 
-        yaml_text = linesep.join(model_content.splitlines()[content_start_line:content_end_line + end_of_file_offset])
+        yaml_text = linesep.join(model_content.splitlines()[content_start_line: content_end_line + end_of_file_offset])
```
Comment (cvchaparro) [inline python/src/aac/io/parser/_parse_source.py:107]: I've had this kind of change cause linter warnings. How were you able to get around that?

File: python/src/aac/lang/definitions/metadata_structure.py [hunk 8]
```diff
@@ -0,0 +1,83 @@
+"""This module provides functions to collate and build a meta structure for an AaC definition."""
+from copy import deepcopy
+from typing import Any, Optional
+from aac.lang.definitions.definition import Definition
+from aac.lang.definitions.lexeme import Lexeme
+from aac.lang.definitions.schema import get_root_schema_definitions, get_schema_for_field
+from aac.lang.definitions.structural_node import NodeType, StructuralNode
+from aac.lang.language_context import LanguageContext
```
Comment (cvchaparro) [inline python/src/aac/lang/definitions/metadata_structure.py:8]:
```suggestion
"""This module provides functions to collate and build a meta structure for an AaC definition."""

from copy import deepcopy
from typing import Any, Optional

from aac.lang.definitions.definition import Definition
from aac.lang.definitions.lexeme import Lexeme
from aac.lang.definitions.schema import get_root_schema_definitions, get_schema_for_field
from aac.lang.definitions.structural_node import NodeType, StructuralNode
from aac.lang.language_context import LanguageContext
```

File: python/src/aac/lang/definitions/metadata_structure.py [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+
+
+def compute_metadata_structure(definition: Definition, language_context: LanguageContext) -> None:
+    """
+    Builds the metadata structure for the definition and attaches it the passed-in definition's metastructure attribute.
```
Comment (cvchaparro) [inline python/src/aac/lang/definitions/metadata_structure.py:13]:
```suggestion
    Builds the metadata structure for the definition and attaches it to the passed-in definition's `meta_structure` attribute.
```

File: python/src/aac/lang/definitions/metadata_structure.py [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+
+    Args:
+        definition (Definition): The definition to compute a metastructure for
+        language_context (LanguageContext): The language context used to provide the metadata
```
Comment (cvchaparro) [inline python/src/aac/lang/definitions/metadata_structure.py:17]:
```suggestion
        definition (Definition): The definition for which to compute a meta-structure.
        language_context (LanguageContext): The language context used to provide the metadata.
```

File: python/src/aac/lang/definitions/metadata_structure.py [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+    root_node = StructuralNode(NodeType.ROOT_KEY, root_definition, root_token, None, [])
+    root_yaml_structure = yaml_traversal_structure.get(root_token.value, {})
+    _recursive_meta_structure_build(
+        definition, token_traversal_list, root_yaml_structure, root_node, [root_token.value], language_context
+    )
+
+    definition.meta_structure = root_node
+
+
+def _recursive_meta_structure_build(
```
Comment (cvchaparro) [inline python/src/aac/lang/definitions/metadata_structure.py:34]: I recommend renaming this to `_build_recursive_meta_structure`

File: python/src/aac/lang/definitions/metadata_structure.py [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+    definition: Definition,
+    tokens: list[Lexeme],
+    yaml_structure: dict[str, Any],
+    parent_node: StructuralNode,
+    traversed_keys: list[str],
+    language_context: LanguageContext,
+) -> None:
+    """Recursively builds the meta structure nodes and attaches them to the parent node."""
```
Comment (cvchaparro) [inline python/src/aac/lang/definitions/metadata_structure.py:42]:
```suggestion
    """Recursively builds the meta-structure nodes and attaches them to the parent node."""
```

File: python/src/aac/lang/definitions/metadata_structure.py (hunk 6 above)
Comment (cvchaparro) [reply python/src/aac/lang/definitions/metadata_structure.py:47]: This is formatted based using our [auto-formatter](https://black.readthedocs.io/en/stable/).

File: python/src/aac/lang/definitions/metadata_structure.py [hunk 13]
```diff
@@ -0,0 +1,83 @@
...
+        new_node = StructuralNode(node_type, defining_definition, node_token, parent_node, [])
+        child_nodes.append(new_node)
+        return new_node
+
+    def _get_node_token_keys_and_definition() -> tuple[Lexeme, list[str], Optional[Definition]]:
+        node_token = tokens.pop(0)
+        node_value = node_token.value
+        node_traversal_keys = [*traversed_keys, node_value]
+        node_definition = get_schema_for_field(definition, node_traversal_keys, language_context)
+        return (node_token, node_traversal_keys, node_definition)
```
Comment (cvchaparro) [inline python/src/aac/lang/definitions/metadata_structure.py:58]: Technically, the parens aren't needed here.

File: python/src/aac/lang/definitions/schema.py [hunk 14]
```diff
@@ -94,8 +102,72 @@ def _get_sub_definitions(schema_definition: Definition, fields):
...
+                field_schema_definition = context.get_enum_definition_by_type(field_type)
+
+        else:
+            # In the case of no field-type, we're assumed to be at a terminal value (enum or primitive value)
+            field_schema_definition = context.get_enum_definition_by_type(key_to_traverse)
+
+        return field_schema_definition
+
+    if len(keys_to_traverse) > 1:
+        keys_to_traverse.pop(0)  # Go ahead and pop the root key since we don't specifically need it.
```
Comment (cvchaparro) [inline python/src/aac/lang/definitions/schema.py:163]: Why have users provide the root key if we don't need it?

File: python/src/aac/lang/definitions/schema.py [hunk 15]
```diff
@@ -94,8 +102,72 @@ def _get_sub_definitions(schema_definition: Definition, fields):
...
+        definition_to_return = _traverse_key(get_schema_defined_fields(source_definition, context))
+    elif len(keys_to_traverse) > 0 and keys_to_traverse[0] == source_definition.get_root_key():
+        definition_to_return = get_definition_schema(source_definition, context)
+
+    return definition_to_return
+
+
+def _convert_fields_list_to_dict(fields: list[dict]) -> dict[str, dict]:
```
Comment (cvchaparro) [inline python/src/aac/lang/definitions/schema.py:171]: I can't place it off hand but don't we have (effectively) this function elsewhere in the code base?

File: python/src/aac/lang/definitions/structural_node.py [hunk 16]
```diff
@@ -0,0 +1,60 @@
...
+    SCHEMA_STRUCTURE = auto()
+    ENUM_VALUE = auto()
+    ROOT_KEY = auto()
+    LIST_ENTRY = auto()
+    PRIMITIVE_VALUE = auto()
+
+
+@attrs(hash=False, eq=False)
+class StructuralNode:
+    """An Architecture-as-Code definition structural node.
```
Comment (cvchaparro) [inline python/src/aac/lang/definitions/structural_node.py:27]:
```suggestion
    """
    An Architecture-as-Code definition structural node.
```

File: python/src/aac/lang/definitions/structural_node.py [hunk 17]
```diff
@@ -0,0 +1,60 @@
...
+
+    These nodes are used to build a traversable tree for the definition's structure as opposed to leveraging the
+        unstructured dictionaries, lists, and primitives supplied by just parsing the yaml structure.
+
+    Attributes:
+        node_type (NodeType): Enum value used to quickly distinguish between node types
```
Comment (cvchaparro) [inline python/src/aac/lang/definitions/structural_node.py:33]:
```suggestion
        node_type (NodeType): Enum value used to quickly distinguish between node types.
```

File: python/src/aac/lang/definitions/structural_node.py (hunk 4 above)
Comment (cvchaparro) [reply python/src/aac/lang/definitions/structural_node.py:34]: I think Lizz is correct - this should be:

```suggestion
        defining_definition (Optional[Definition]): The `Definition` that defines the structure or enumeration values for the token.
```

File: python/src/aac/lang/definitions/structural_node.py [hunk 18]
```diff
@@ -0,0 +1,60 @@
...
+        defining_definition (Definition): The definition that defines the structure or enumeration values for the token.
+        token (Optional[Lexeme]): The corresponding token within the definition's yaml structure. Optional as nodes representing entries in a list will not have tokens.
+        parent (Optional[StructuralNode]): Reference to the node's parent; null if the node is the root node.
```
Comment (cvchaparro) [inline python/src/aac/lang/definitions/structural_node.py:36]:
```suggestion
        parent (Optional[StructuralNode]): Reference to the node's parent; None if the node is the root node.
```

File: python/src/aac/lang/definitions/structural_node.py (hunk 5 above)
Comment (cvchaparro) [reply python/src/aac/lang/definitions/structural_node.py:40]: I think the comment addresses the your question, @lizzcondrey.

File: python/src/aac/lang/definitions/structural_node.py [hunk 19]
```diff
@@ -0,0 +1,60 @@
...
+        children (List[StructuralNode]): Reference the the node's children; empty if there are no children.
+    """
+    node_type: NodeType = attrib(validator=validators.instance_of(NodeType))
+    defining_definition: Optional['Definition'] = attrib()  # Can't reference Definition here for validation without creating a circular dependency.
+    token: Optional[Lexeme] = attrib(validator=validators.optional(validators.instance_of(Lexeme)))
```
Comment (cvchaparro) [inline python/src/aac/lang/definitions/structural_node.py:41]: It's probably time for us to rename `Lexeme` to `Token`.

File: python/src/aac/lang/definitions/structural_node.py [hunk 20]
```diff
@@ -0,0 +1,60 @@
...
+    parent: Optional[StructuralNode] = attrib()  # Python/Attrs isn't capable of supporting self-referencing types for self-referential type validator
```
Comment (cvchaparro) [inline python/src/aac/lang/definitions/structural_node.py:42]: For self-referential validators we can make use of [custom validators](https://www.attrs.org/en/stable/examples.html#validators) for this. Something like this should work:

```
    parent: Optional[StructuralNode] = attrib()
    @parent.validator
    def _validate_is_structural_node(self, attribute, value):
        if value is not None or not isinstance(value, StructuralNode):
            raise ValueError(f"{attribute} is not valid!")
```

File: python/src/aac/lang/definitions/structural_node.py [hunk 21]
```diff
@@ -0,0 +1,60 @@
...
+    children: list[StructuralNode] = attrib(validator=validators.instance_of(list))
+
+    def __init__(self, node_type: NodeType, defining_definition: Optional['Definition'] = None, token: Optional[Lexeme] = None, parent: Optional[StructuralNode] = None, children: list[StructuralNode] = []):
+        """
+        Init the Structural Node Class.
```
Comment (cvchaparro) [inline python/src/aac/lang/definitions/structural_node.py:47]:
```suggestion
        Initialize a Structural Node object.
```

File: python/src/aac/plugins/first_party/gen_design_doc/gen_design_doc_impl.py
Comment (cvchaparro) [inline python/src/aac/plugins/first_party/gen_design_doc/gen_design_doc_impl.py:1]: Generally, I'd agree we should do this but I'd suggest we revert these changes since I had to make substantive changes to `gen_design_doc_impl.py` in my PR (#500) and, as part of that one, I also made these changes? If we keep this as-is, it'll cause some merge issues that can lose some of the changes I made in my PR.

**Review by cvchaparro — CHANGES_REQUESTED** [review]

File: python/src/aac/lang/definitions/metadata_structure.py (hunk 6 above)
Comment (lizzcondrey) [reply python/src/aac/lang/definitions/metadata_structure.py:47]: Ah, so guessing that has a threshold for swapping to single line entries then and this hadn't met it.

File: python/src/aac/lang/definitions/definition.py [hunk 22]
```diff
@@ -45,6 +48,9 @@ class Definition:
     lexemes: list[Lexeme] = attrib(default=Factory(list), validator=validators.instance_of(list))
     structure: dict = attrib(default=Factory(dict), validator=validators.instance_of(dict))
     imports: list[str] = attrib(default=Factory(list), validator=validators.instance_of(list))
+    meta_structure: Optional[StructuralNode] = attrib(
+        default=None, validator=validators.optional(validators.instance_of(StructuralNode))
```
[the remark this answers is not in the document]
Comment (Coffee2Bits) [reply python/src/aac/lang/definitions/definition.py:52]: We're using the `attrs` library to make our Python experience a dash easier. This is the [link](https://www.attrs.org/en/stable/examples.html#validators) to the docs, pointing at the validators. 

These attrs validators just enforce that the values put in those fields adhere to the type constraints defined by the attrs validators. In the lines you highlighted, I have an instance check for StructuralNode, but without the optional wrapper we'll get runtime errors when trying to assign any value other than an instance of StructuralNode, like None, to the field.

File: python/src/aac/lang/definitions/metadata_structure.py (hunk 6 above)
Comment (Coffee2Bits) [reply python/src/aac/lang/definitions/metadata_structure.py:47]: Yeah, it's a few characters shy of taking it to the extreme of one-line per parameter.

File: python/src/aac/io/parser/_parse_source.py (hunk 7 above)
Comment (Coffee2Bits) [reply python/src/aac/io/parser/_parse_source.py:107]: I... I don't know. I wonder if a rule was disabled.

File: python/src/aac/lang/definitions/structural_node.py (hunk 20 above)
Comment (Coffee2Bits) [reply python/src/aac/lang/definitions/structural_node.py:42]: Sweet find! That'll be nice.

File: python/src/aac/lang/definitions/metadata_structure.py [hunk 23]
```diff
@@ -0,0 +1,83 @@
...
+) -> None:
+    """Recursively builds the meta structure nodes and attaches them to the parent node."""
+
+    child_nodes = []
+
+    def _create_node(
+        node_type: NodeType, node_token: Optional[Lexeme] = None, defining_definition: Optional[Definition] = None
+    ) -> StructuralNode:
+        new_node = StructuralNode(node_type, defining_definition, node_token, parent_node, [])
+        child_nodes.append(new_node)
```
[the remark this answers is not in the document]
Comment (Coffee2Bits) [reply python/src/aac/lang/definitions/metadata_structure.py:52]: Yep, you've got it.

File: python/src/aac/lang/definitions/metadata_structure.py (hunk 13 above)
Comment (Coffee2Bits) [reply python/src/aac/lang/definitions/metadata_structure.py:60]: I can't find a consensus on style. I guess if it's not functionally different, then I'd prefer less.
```suggestion
        return node_token, node_traversal_keys, node_definition
```

File: python/src/aac/lang/definitions/schema.py (hunk 14 above)
Comment (Coffee2Bits) [reply python/src/aac/lang/definitions/schema.py:163]: You do need it on line 165, just not in the recursive calls.

File: python/src/aac/lang/definitions/schema.py (hunk 15 above)
Comment (Coffee2Bits) [reply python/src/aac/lang/definitions/schema.py:171]: This is a common pattern we have in some areas, but it is just a dictionary comprehension.

File: python/src/aac/lang/definitions/structural_node.py (hunk 5 above)
Comment (Coffee2Bits) [reply python/src/aac/lang/definitions/structural_node.py:41]:
```suggestion
    defining_definition: Optional[Definition] = attrib()  # Can't reference Definition here for validation without creating a circular dependency.
```

File: python/src/aac/plugins/first_party/gen_design_doc/gen_design_doc_impl.py
Comment (Coffee2Bits) [reply python/src/aac/plugins/first_party/gen_design_doc/gen_design_doc_impl.py:1]: reverted

File: python/src/aac/io/parser/_parse_source.py (hunk 7 above)
Comment (cvchaparro) [reply python/src/aac/io/parser/_parse_source.py:107]: No, I figured it out (although it's a bit weird). `flake8` is configured to allow space after a colon (which makes sense for - e.g. `lambda: None`, etc) but not before one and the [E203](https://www.flake8rules.com/rules/E203.html) error - i.e. "whitespace before ':'" - is what I was seeing.

File: python/src/aac/lang/definitions/structural_node.py (hunk 19 above)
Comment (cvchaparro) [reply python/src/aac/lang/definitions/structural_node.py:42]: To clarify, I think we do this in a separate PR.

**Review by cvchaparro — APPROVED** [review]
Nice!
